experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+//! }
+//!
+//! impl<'a> Variants for &'a OwnedVariant {
+//!     type Variant = &'a String;
+//!     type Iter = std::slice::Iter<'a, String>;
+//!     fn variants(self) -> Self::Iter {
+//!         self.variants.iter()
+//!     }
+//! }
+//!
+//! let owned = OwnedVariant{ variants: vec!["a".to_string(), "b".to_string()], };
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     owned.variants()
+//!         .map(|v| v.to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
+//! ```
+//!
+
+/// This trait contains the common features of the Locale object that must be shared among
+/// all the implementations.  Every implementor of `listformat` should provide their
+/// own version of [Locale], and should ensure that it implements [Locale]. as
+/// specified here.
+///
+/// A stronger type definition is also possible, pending agreement.  Until no firm implementations
+/// exist, we reserve the option to strenghten the contract of [Locale] as we see fit.
+///
+/// For example, here are some possible methods that this trait could implement instead:
+///
+/// - `language`
+/// - `script`
+/// - `variant`
+/// - (...)
+pub trait Locale<'a> {
+    /// Gives the full representation of the `Locale` in the [BCP 47][bcp47] string encoding.
+    ///
+    ///   [bcp47]: https://tools.ietf.org/html/bcp47
+    fn to_bcp47() -> &'a str;
+}
+
+/// The [listformat] mod contains all the needed implementation bits for `Intl.ListFormat`.
+///
+/// > Note: This is not yet the entire API.  I'd like to get to a consensus on what has been
+/// defined, then use the patterns adopted here for the rest.
+pub mod listformat {
+    /// Contains the API configuration as prescribed by ECMA 402.
+    ///
+    /// The meaning of the options is the same as in the similarly named
+    /// options in the JS version.
+    ///
+    /// See [options::In] for the contents of the options.  See the [Factory]
+    /// trait for the use of the options.
+    pub mod options {
+        /// Chooses the list formatting approach.
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Style {
+            Long,
+            Short,
+            Narrow,
+        }
+        /// Chooses between "this, that and other", and "this, that or other".
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Type {
+            /// "This, that and other".
+            Conjunction,
+            /// "This, that or other".
+            Disjunction,
+        }
+
+        /// The options set by the user at construction time.  See discussion at the top level
+        /// about the name choice.  Provides as a "bag of options" since we don't expect any
+        /// implementations to be attached to this struct.
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub struct In {
+            /// Selects a [Style] for the formatted list.
+            pub style: Style,
+            /// Selects a [Type] for the formatted list.
+            pub in_type: Type,
+        }
```

Comment:
I'd consider naming it `ListFormatOptions`.

Reply:
And implement `Default` for it, so that user can then cal `ListFormat::try_new(locale, ListFormatOptions::default())` :)

Reply:
Also, `Default` supports constructs such as:

```
let options = ListFormatOptions {
  style: Style::Long,
  ..Default::default()
};
```